include/Dialect/LWE/IR/LWEAttributes.td

```diff
@@ -55,10 +67,143 @@ def LWE_BitFieldEncoding: LWE_EncodingAttr<"BitFieldEncoding", "bit_field_encodi
+  let summary = "An attribute describing encoded RLWE plaintexts via evaluations at fixed points.";
+  let description = [{
+    A "evaluation encoding" of a list of integers $(v_1, \dots, v_n)$ asserts
+    that $f(\x_1) = v_1, \dots, f(x_n) = v_n$ for some implicit, but fixed and
+    distinct choice of inputs $x_i$. The encoded values are also scaled by a
+    scale factor, having the same semantics as `bit_field_encoding`, but
+    applied entry-wise (to either the coefficient or evaluation representation).
+
+    This attribute can be used in multiple ways:
+
+    - On a `poly.poly`, it asserts that the polynomial has been transformed
+      from an evaluation tensor.
+    - On a tensor of `poly.poly`, it asserts that the tensor is an RLWE
+      ciphertext for some RLWE scheme that supports the evaluation encoding.
+
+    A typical workflow for the BFV/BGV schemes using this encoding would be
+    to apply a INTT operation to the input list of cleartexts to convert from
+    evaluation form to coefficient form, then encrypt the resulting polynomial
+    in coefficient form, then apply NTT back to the evaluation form for faster
+    multiplication of ciphertexts.
+
+    The points chosen are fixed to be the powers of a primitive root of unity
+    in the relevant coefficient domain, which allows one to use NTT/INTT to
+    tansform quickly between the coefficient and evaluation forms.
+
+    Example:
+
+    ```
+    #generator = #poly.polynomial<1 + x**1024>
+    #ring = #poly.ring<cmod=65536, ideal=#generator>
+    #lwe_encoding = #lwe.poly_evaluation_encoding<cleartext_start=30, cleartext_bitwidth=3>
+
+    %evals = arith.constant <[1, 2, 4, 5]> : tensor<4xi16>
+    // Note no `intt` operation exists in poly yet.
+    %poly1 = poly.intt %evals : tensor<4xi16> -> !poly.poly<#ring, #eval_encoding>
+    %poly2 = poly.intt %evals : tensor<4xi16> -> !poly.poly<#ring, #eval_encoding>
+    %rlwe_ciphertext = tensor.from_elements %poly1, %poly2 : tensor<2x!poly.poly<#ring, #eval_encoding>>
+    ```
+  }];
 }
+
+def RLWE_PolyRoundedEmbeddingEncoding
+  : LWE_EncodingAttrWithScalingFactor<"PolyRoundedEmbeddingEncoding", "poly_rounded_embedding_encoding"> {
+  let summary = "An attribute describing encoded RLWE plaintexts via the rounded inverse canonical embedding.";
+  let description = [{
+    Let $n$ be the degree of the polynomials in the plaintext space.
+    A "rounded embedding encoding" of a list of real or complex values $v_1,
+    \dots, v_{n/2}$ is (almost) the inverse of the following decoding map.
+
+    Map a polynomial $p \in \mathbb{Z}[x] / (x^N + 1) \to \mathbb{C}^{N/2}$
+    by evaluating it at the following $N/2$ points, where
+    $\omega = e^{2 \pi i / 2N}$ is the primitive $2N$th root of unity:
+
+    \[
+      \omega, \omega^3, \omega^5, \dots, \omega^{N-1}
+    \]
+
+    This map is called the "canonical embedding" $\tau$, though conjugates are
+    discarded for efficiency because $p(\overline{z}) = \overline{p(z)}$.
```

Comment:
z has an imaginary part if and only if conj(z) does, so throwing out conjugates is not enough to ensure the inverse is real.

I think what you're saying is that the cleartext space would not be allowed to be all of C^{N} (if you hypothetically used the full canonical embedding), but rather the subspace consisting of vectors (x_1, \dots, x_N) for which x_i = conj(x_{N-i}). That's the property that makes the inverse real, and those conditions reduce the dimension by half when N is a power of 2 (another mistake in the above is not assuming a power of 2 N; in general I seem to remember the new dimension is euler's totient of N?)

I need to recall the proof of that property ensuring the inverse is real. Do you have a reference to a proof?

If that's right, then the reason is still kind of efficiency: we don't represent the entire N-dimensional codomain with conjugates included because the second half is duplicate information enforced by those constraints (and the constraints "come first" semantically).

Reply:
Er-- wait, based on what I said throwing out conjugates IS enough to make it real :)

I think I can reword it for clarity anyway.

Reply:
It follows from the fundamental theorem of algebra https://en.wikipedia.org/wiki/Fundamental_theorem_of_algebra and https://en.wikipedia.org/wiki/Complex_conjugate_root_theorem.

Reply:
Let's chat about it tomorrow.